File: internal/crypto/content.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Serraniel and the Sendan contributors

package crypto

import (
	"crypto/cipher"
	"crypto/hkdf"
	"crypto/sha256"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
)

// Content encoding parameters, per spec §5.
const (
	// RecordSize is the fixed record size in bytes.
	RecordSize = 65536

	// ContentSaltSize is the length of the per-upload content salt.
	ContentSaltSize = 16

	// headerSize is cSalt(16) || rs(4) || idlen(1), with an empty keyid.
	headerSize = ContentSaltSize + 4 + 1

	// maxRecordPlaintext is the record size less the GCM tag and the delimiter.
	maxRecordPlaintext = RecordSize - 16 - 1

	delimiterNonFinal = 0x01
	delimiterFinal    = 0x02

	// maxSequence bounds the record counter well below the point at which the
	// 96-bit nonce space could wrap. Exceeding it would risk nonce reuse, which
	// discloses the GCM authentication key.
	maxSequence = 1 << 48
)

// Info strings for content key derivation, per spec §5.1.
//
// The aes256gcm designation is deliberate: RFC 8188 defines only aes128gcm, and
// Sendan uses the same framing with a 256-bit key. A distinct info string keeps
// the two from ever deriving the same key material.
var (
	infoContentKey   = "Content-Encoding: aes256gcm\x00"
	infoContentNonce = "Content-Encoding: nonce\x00"
)

// ErrContent reports a malformed, truncated, or tampered content stream.
var ErrContent = errors.New("crypto: invalid content stream")

// deriveContentKeys produces the content encryption key and nonce base from the
// file key and the content salt.
func deriveContentKeys(fileKey, contentSalt []byte) (cek []byte, nonceBase [12]byte, err error) {
	if len(fileKey) != FileKeySize {
		return nil, nonceBase, fmt.Errorf("%w: file key is %d bytes, want %d", ErrKeyMaterial, len(fileKey), FileKeySize)
	}
	if len(contentSalt) != ContentSaltSize {
		return nil, nonceBase, fmt.Errorf("%w: content salt is %d bytes, want %d", ErrKeyMaterial, len(contentSalt), ContentSaltSize)
	}

	prk, err := hkdf.Extract(sha256.New, fileKey, contentSalt)
	if err != nil {
		return nil, nonceBase, fmt.Errorf("crypto: hkdf extract: %w", err)
	}
	cek, err = hkdf.Expand(sha256.New, prk, infoContentKey, 32)
	if err != nil {
		return nil, nonceBase, fmt.Errorf("crypto: hkdf expand content key: %w", err)
	}
	nb, err := hkdf.Expand(sha256.New, prk, infoContentNonce, 12)
	if err != nil {
		return nil, nonceBase, fmt.Errorf("crypto: hkdf expand nonce: %w", err)
	}
	copy(nonceBase[:], nb)
	return cek, nonceBase, nil
}

// recordNonce derives the nonce for a record, per spec §5.3:
// the nonce base exclusive-ORed with the big-endian record sequence number.
//
// The sequence number is a strict counter and is never random. Two records of
// one stream must never share a nonce; see the warning in spec §5.3.
func recordNonce(base [12]byte, seq uint64) []byte {
	n := base
	// The counter is 96-bit big-endian, but a uint64 sequence can only occupy
	// the low eight octets, so the leading four are always zero.
	var counter [8]byte
	binary.BigEndian.PutUint64(counter[:], seq)
	for i, b := range counter {
		n[4+i] ^= b
	}
	return n[:]
}

// recordSealer owns the record counter and is the only way to encrypt a record.
//
// This exists to make nonce reuse structurally impossible rather than merely
// discouraged. AES-GCM nonce reuse under one key discloses the authentication
// key, not merely one message, so the invariant is worth enforcing in the type
// system rather than in a comment: the nonce cannot be supplied by a caller,
// and every call to seal advances the counter before returning.
//
// Decryption deliberately has no equivalent. Reusing a nonce to *open* a record
// is not a hazard, and the decryptor must be able to reject an out-of-order
// record rather than silently advance past it.
type recordSealer struct {
	aead      cipher.AEAD
	nonceBase [12]byte
	seq       uint64
}

func (s *recordSealer) seal(plaintext []byte) ([]byte, error) {
	if s.seq >= maxSequence {
		return nil, fmt.Errorf("%w: record sequence exhausted", ErrContent)
	}
	// The counter advances unconditionally, including on the path where Seal
	// would panic, so no sequence value can ever be issued twice.
	seq := s.seq
	s.seq++
	return s.aead.Seal(nil, recordNonce(s.nonceBase, seq), plaintext, nil), nil
}

// Encryptor encrypts a byte stream into the Sendan content encoding.
//
// Close must be called and its error checked: the final record carries the
// terminating delimiter, and without it a decryptor correctly treats the stream
// as truncated.
type Encryptor struct {
	w             io.Writer
	sealer        *recordSealer
	pendingHeader []byte
	buf           []byte
	n             int
	headerWritten bool
	closed        bool
	err           error
}

// NewEncryptor returns an [Encryptor] writing to w.
//
// It generates a random content salt, which is emitted in the header.
func NewEncryptor(w io.Writer, fileKey []byte) (*Encryptor, error) {
	salt, err := randomBytes(ContentSaltSize)
	if err != nil {
		return nil, err
	}
	return newEncryptorWithSalt(w, fileKey, salt)
}

// newEncryptorWithSalt allows tests and the vector generator to pin the salt.
// Callers outside this package must use [NewEncryptor] so the salt is random.
func newEncryptorWithSalt(w io.Writer, fileKey, contentSalt []byte) (*Encryptor, error) {
	cek, nonceBase, err := deriveContentKeys(fileKey, contentSalt)
	if err != nil {
		return nil, err
	}
	aead, err := newAEAD(cek)
	if err != nil {
		return nil, err
	}

	header := make([]byte, 0, headerSize)
	header = append(header, contentSalt...)
	header = binary.BigEndian.AppendUint32(header, RecordSize)
	header = append(header, 0) // idlen: the keyid is always empty

	return &Encryptor{
		w:             w,
		sealer:        &recordSealer{aead: aead, nonceBase: nonceBase},
		pendingHeader: header,
		buf:           make([]byte, maxRecordPlaintext),
	}, nil
}

// writeHeader emits the header lazily, so constructing an Encryptor performs
// no I/O.
func (e *Encryptor) writeHeader() error {
	if e.headerWritten {
		return nil
	}
	if _, err := e.w.Write(e.pendingHeader); err != nil {
		return fmt.Errorf("crypto: write header: %w", err)
	}
	e.headerWritten = true
	return nil
}

// Write buffers plaintext, emitting a record whenever a full one accumulates
// and further data is known to follow.
//
// A full buffer is never emitted eagerly, because whether it is the final
// record is not known until either more data arrives or Close is called.
func (e *Encryptor) Write(p []byte) (int, error) {
	if e.err != nil {
		return 0, e.err
	}
	if e.closed {
		return 0, errors.New("crypto: write after close")
	}

	written := 0
	for len(p) > 0 {
		if e.n == maxRecordPlaintext {
			// The buffer is full and more data remains, so this record is
			// certainly not the final one.
			if err := e.flush(false); err != nil {
				e.err = err
				return written, err
			}
		}
		k := copy(e.buf[e.n:], p)
		e.n += k
		p = p[k:]
		written += k
	}
	return written, nil
}

// Close emits the final record and must be called exactly once.
func (e *Encryptor) Close() error {
	if e.err != nil {
		return e.err
	}
	if e.closed {
		return nil
	}
	e.closed = true
	if err := e.flush(true); err != nil {
		e.err = err
		return err
	}
	return nil
}

func (e *Encryptor) flush(final bool) error {
	if err := e.writeHeader(); err != nil {
		return err
	}

	delimiter := byte(delimiterNonFinal)
	if final {
		delimiter = delimiterFinal
	}

	plaintext := make([]byte, e.n+1)
	copy(plaintext, e.buf[:e.n])
	plaintext[e.n] = delimiter

	record, err := e.sealer.seal(plaintext)
	if err != nil {
		return err
	}
	if _, err := e.w.Write(record); err != nil {
		return fmt.Errorf("crypto: write record: %w", err)
	}
	e.n = 0
	return nil
}

// Decryptor decrypts a stream produced by [Encryptor].
type Decryptor struct {
	r         io.Reader
	aead      cipher.AEAD
	nonceBase [12]byte
	seq       uint64
	fileKey   []byte

	plain      []byte
	off        int
	headerRead bool
	sawFinal   bool
	err        error
	record     []byte
}

// NewDecryptor returns a [Decryptor] reading from r.
//
// The header is parsed on the first Read, so constructing one performs no I/O.
func NewDecryptor(r io.Reader, fileKey []byte) (*Decryptor, error) {
	if len(fileKey) != FileKeySize {
		return nil, fmt.Errorf("%w: file key is %d bytes, want %d", ErrKeyMaterial, len(fileKey), FileKeySize)
	}
	return &Decryptor{r: r, fileKey: fileKey, record: make([]byte, RecordSize)}, nil
}

func (d *Decryptor) readHeader() error {
	header := make([]byte, headerSize)
	if _, err := io.ReadFull(d.r, header); err != nil {
		return ErrContent
	}

	salt := header[:ContentSaltSize]
	rs := binary.BigEndian.Uint32(header[ContentSaltSize : ContentSaltSize+4])
	idlen := header[ContentSaltSize+4]

	// The record size and key identifier are fixed by the specification. They
	// are validated rather than honoured: accepting a value from the stream
	// would be a negotiated parameter, which spec §11 forbids.
	if rs != RecordSize || idlen != 0 {
		return ErrContent
	}

	cek, nonceBase, err := deriveContentKeys(d.fileKey, salt)
	if err != nil {
		return ErrContent
	}
	aead, err := newAEAD(cek)
	if err != nil {
		return ErrContent
	}
	d.aead = aead
	d.nonceBase = nonceBase
	d.headerRead = true
	return nil
}

// Read implements [io.Reader].
//
// A stream that ends without a final record yields [ErrContent] rather than a
// short read, so truncation can never be mistaken for a complete file.
func (d *Decryptor) Read(p []byte) (int, error) {
	if d.err != nil {
		return 0, d.err
	}
	if !d.headerRead {
		if err := d.readHeader(); err != nil {
			d.err = err
			return 0, err
		}
	}

	for d.off == len(d.plain) {
		if d.sawFinal {
			d.err = io.EOF
			return 0, io.EOF
		}
		if err := d.nextRecord(); err != nil {
			d.err = err
			return 0, err
		}
	}

	n := copy(p, d.plain[d.off:])
	d.off += n
	return n, nil
}

func (d *Decryptor) nextRecord() error {
	if d.seq >= maxSequence {
		return ErrContent
	}

	n, err := io.ReadFull(d.r, d.record)
	switch {
	case err == nil:
	case errors.Is(err, io.ErrUnexpectedEOF):
		// A short record is permitted only as the last one in the stream.
	case errors.Is(err, io.EOF):
		// The stream ended without a record carrying the final delimiter.
		return ErrContent
	default:
		return ErrContent
	}

	plaintext, err := d.aead.Open(nil, recordNonce(d.nonceBase, d.seq), d.record[:n], nil)
	if err != nil {
		return ErrContent
	}
	// Spec §5.4: a record with no plaintext has no delimiter to inspect. This
	// is load-bearing rather than defensive - the delimiter is read as
	// plaintext[len(plaintext)-1] below, and on an empty slice that panics
	// rather than returning something the switch can reject. The TypeScript
	// implementation keeps the same check for the same rule, where an
	// out-of-range read yields undefined and the delimiter switch would catch
	// it anyway.
	if len(plaintext) == 0 {
		return ErrContent
	}

	// The delimiter is the final octet. Optional zero padding, which RFC 8188
	// permits, is deliberately not accepted: it would make two distinct
	// encodings of one plaintext valid, and malleability of any kind is not
	// worth the flexibility here.
	delimiter := plaintext[len(plaintext)-1]
	switch delimiter {
	case delimiterFinal:
		d.sawFinal = true
	case delimiterNonFinal:
		// A non-final record must be full; otherwise a truncated stream could
		// be presented as a complete one.
		if n != RecordSize {
			return ErrContent
		}
	default:
		return ErrContent
	}

	d.plain = plaintext[:len(plaintext)-1]
	d.off = 0
	d.seq++

	if d.sawFinal {
		// Nothing may follow the final record.
		var probe [1]byte
		if extra, _ := d.r.Read(probe[:]); extra > 0 {
			return ErrContent
		}
	}
	return nil
}

// EncodedContentLength reports the encoded size of a plaintext of n bytes.
//
// An uploader must declare the total length before sending the first byte, and
// what it sends is the encoding rather than the file, so it has to know this
// without having produced it. Being wrong is not a rounding error: too small
// and the server refuses the tail, too large and the upload never completes.
//
// The encoding is fully determined by the length, which is what makes this
// possible. Records hold maxRecordPlaintext bytes; the final one is short, and
// is emitted even for an empty input because the terminating delimiter is what
// distinguishes a complete stream from a truncated one.
//
// This is the Go half of a value the TypeScript client computes too. The two
// must agree exactly: a browser and the command line client uploading the same
// file declare the same number, and a shared test vector pins them together.
func EncodedContentLength(n int64) (int64, error) {
	if n < 0 {
		return 0, fmt.Errorf("%w: plaintext length %d is not a byte count", ErrContent, n)
	}

	// A whole multiple of the record size fills its last record exactly, and
	// that record is the final one. Rounding up would claim a further empty
	// record that the encoder never emits.
	var nonFinal int64
	if n > 0 {
		nonFinal = (n + maxRecordPlaintext - 1) / maxRecordPlaintext
		nonFinal--
	}
	final := n - nonFinal*maxRecordPlaintext

	// The delimiter and the GCM tag are what a record costs beyond its
	// plaintext.
	const recordOverhead = 1 + 16
	return headerSize + nonFinal*RecordSize + final + recordOverhead, nil
}

File: web/src/crypto/content.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Serraniel and the Sendan contributors

import { ContentError, KeyMaterialError } from "./errors.js";
import { FILE_KEY_SIZE, randomBytes } from "./keys.js";

/** Content encoding parameters, per spec §5. */
export const RECORD_SIZE = 65536;
export const CONTENT_SALT_SIZE = 16;

/** cSalt(16) || rs(4) || idlen(1), with an empty keyid. */
export const HEADER_SIZE = CONTENT_SALT_SIZE + 4 + 1;

/** The record size less the GCM tag and the delimiter. */
export const MAX_RECORD_PLAINTEXT = RECORD_SIZE - 16 - 1;

const DELIMITER_NON_FINAL = 0x01;
const DELIMITER_FINAL = 0x02;

/** A record's ciphertext less its plaintext: the delimiter and the GCM tag. */
const RECORD_OVERHEAD = 1 + 16;

/**
 * The encoded length of a plaintext of the given size.
 *
 * An uploader must declare the total length before sending the first byte, and
 * what it sends is the encoding rather than the file, so it has to know this
 * without having produced it. Being wrong is not a rounding error: too small
 * and the server refuses the tail, too large and the upload never completes.
 *
 * The encoding is fully determined by the length, which is what makes this
 * possible. Records hold {@link MAX_RECORD_PLAINTEXT} bytes; the final one is
 * short, and is emitted even for an empty file because the terminating
 * delimiter is what distinguishes a complete stream from a truncated one.
 */
export function encodedContentLength(plaintextLength: number): number {
  if (!Number.isSafeInteger(plaintextLength) || plaintextLength < 0) {
    throw new ContentError(`plaintext length ${plaintextLength} is not a byte count`);
  }
  // A whole multiple of the record size fills its last record exactly, and that
  // record is the final one. Rounding up would claim a further empty record
  // that the encoder never emits.
  const nonFinal =
    plaintextLength === 0 ? 0 : Math.ceil(plaintextLength / MAX_RECORD_PLAINTEXT) - 1;
  const finalPlaintext = plaintextLength - nonFinal * MAX_RECORD_PLAINTEXT;
  return HEADER_SIZE + nonFinal * RECORD_SIZE + finalPlaintext + RECORD_OVERHEAD;
}

/**
 * Bounds the record counter well below the point at which the 96-bit nonce
 * space could wrap. Exceeding it would risk nonce reuse, which discloses the
 * GCM authentication key.
 */
const MAX_SEQUENCE = 2 ** 48;

/**
 * Info strings for content key derivation, per spec §5.1.
 *
 * The aes256gcm designation is deliberate: RFC 8188 defines only aes128gcm, and
 * Sendan uses the same framing with a 256-bit key. A distinct info string keeps
 * the two from ever deriving the same key material.
 *
 * Built with fromCharCode so no literal NUL appears in the source.
 */
const NUL = String.fromCharCode(0);
const INFO_CONTENT_KEY = `Content-Encoding: aes256gcm${NUL}`;
const INFO_CONTENT_NONCE = `Content-Encoding: nonce${NUL}`;

const encoder = new TextEncoder();

interface ContentKeys {
  key: CryptoKey;
  nonceBase: Uint8Array;
}

async function deriveContentKeys(
  fileKey: Uint8Array,
  contentSalt: Uint8Array,
): Promise<ContentKeys> {
  if (fileKey.length !== FILE_KEY_SIZE) {
    throw new KeyMaterialError(`file key is ${fileKey.length} bytes, want ${FILE_KEY_SIZE}`);
  }
  if (contentSalt.length !== CONTENT_SALT_SIZE) {
    throw new KeyMaterialError(
      `content salt is ${contentSalt.length} bytes, want ${CONTENT_SALT_SIZE}`,
    );
  }

  const base = await crypto.subtle.importKey("raw", fileKey as BufferSource, "HKDF", false, [
    "deriveBits",
  ]);
  const derive = async (info: string, bytes: number) =>
    new Uint8Array(
      await crypto.subtle.deriveBits(
        {
          name: "HKDF",
          hash: "SHA-256",
          salt: contentSalt as BufferSource,
          info: encoder.encode(info) as BufferSource,
        },
        base,
        bytes * 8,
      ),
    );

  const [raw, nonceBase] = await Promise.all([
    derive(INFO_CONTENT_KEY, 32),
    derive(INFO_CONTENT_NONCE, 12),
  ]);
  const key = await crypto.subtle.importKey("raw", raw as BufferSource, "AES-GCM", false, [
    "encrypt",
    "decrypt",
  ]);
  return { key, nonceBase };
}

/**
 * The nonce for a record, per spec §5.3: the nonce base exclusive-ORed with the
 * big-endian record sequence number.
 *
 * The sequence number is a strict counter and is never random. Two records of
 * one stream must never share a nonce.
 */
function recordNonce(base: Uint8Array, seq: number): Uint8Array {
  const n = Uint8Array.from(base);
  // The counter is 96-bit big-endian, but a sequence below 2^48 can only
  // occupy the low six octets.
  const view = new DataView(new ArrayBuffer(8));
  view.setBigUint64(0, BigInt(seq));
  for (let i = 0; i < 8; i++) {
    // biome-ignore lint/style/noNonNullAssertion: indices 4..11 are always in range
    n[4 + i] = n[4 + i]! ^ view.getUint8(i);
  }
  return n;
}

function buildHeader(contentSalt: Uint8Array): Uint8Array {
  const header = new Uint8Array(HEADER_SIZE);
  header.set(contentSalt, 0);
  new DataView(header.buffer).setUint32(CONTENT_SALT_SIZE, RECORD_SIZE);
  header[CONTENT_SALT_SIZE + 4] = 0; // idlen: the keyid is always empty
  return header;
}

/**
 * A TransformStream encrypting plaintext into the Sendan content encoding.
 *
 * A buffered full record is never emitted eagerly, because whether it is the
 * final record is unknown until either more data arrives or the stream ends.
 * The final record carries the terminating delimiter, so a consumer that never
 * receives it correctly treats the stream as truncated.
 */
export function encryptStream(
  fileKey: Uint8Array,
  contentSalt: Uint8Array = randomBytes(CONTENT_SALT_SIZE),
): TransformStream<Uint8Array, Uint8Array> {
  let keys: ContentKeys | null = null;
  let seq = 0;
  const buf = new Uint8Array(MAX_RECORD_PLAINTEXT);
  let n = 0;
  let headerWritten = false;

  const flush = async (
    controller: TransformStreamDefaultController<Uint8Array>,
    final: boolean,
  ) => {
    if (!keys) {
      keys = await deriveContentKeys(fileKey, contentSalt);
    }
    if (!headerWritten) {
      controller.enqueue(buildHeader(contentSalt));
      headerWritten = true;
    }
    if (seq >= MAX_SEQUENCE) {
      throw new ContentError("record sequence exhausted");
    }

    const plaintext = new Uint8Array(n + 1);
    plaintext.set(buf.subarray(0, n), 0);
    plaintext[n] = final ? DELIMITER_FINAL : DELIMITER_NON_FINAL;

    const record = new Uint8Array(
      await crypto.subtle.encrypt(
        { name: "AES-GCM", iv: recordNonce(keys.nonceBase, seq) as BufferSource, tagLength: 128 },
        keys.key,
        plaintext as BufferSource,
      ),
    );
    controller.enqueue(record);
    seq++;
    n = 0;
  };

  return new TransformStream<Uint8Array, Uint8Array>({
    async transform(chunk, controller) {
      let offset = 0;
      while (offset < chunk.length) {
        if (n === MAX_RECORD_PLAINTEXT) {
          // The buffer is full and more data remains, so this record is
          // certainly not the final one.
          await flush(controller, false);
        }
        const take = Math.min(MAX_RECORD_PLAINTEXT - n, chunk.length - offset);
        buf.set(chunk.subarray(offset, offset + take), n);
        n += take;
        offset += take;
      }
    },
    async flush(controller) {
      await flush(controller, true);
    },
  });
}

/**
 * A TransformStream decrypting the Sendan content encoding.
 *
 * A stream that ends without a final record raises {@link ContentError} rather
 * than completing, so truncation can never be mistaken for a complete file.
 */
export function decryptStream(fileKey: Uint8Array): TransformStream<Uint8Array, Uint8Array> {
  let keys: ContentKeys | null = null;
  let seq = 0;
  let sawFinal = false;
  let pending = new Uint8Array(0);

  const append = (chunk: Uint8Array) => {
    const next = new Uint8Array(pending.length + chunk.length);
    next.set(pending, 0);
    next.set(chunk, pending.length);
    pending = next;
  };

  const readHeader = async () => {
    const salt = pending.subarray(0, CONTENT_SALT_SIZE);
    const view = new DataView(pending.buffer, pending.byteOffset, pending.byteLength);
    const rs = view.getUint32(CONTENT_SALT_SIZE);
    const idlen = pending[CONTENT_SALT_SIZE + 4];

    // The record size and key identifier are fixed by the specification. They
    // are validated rather than honoured: accepting a value from the stream
    // would be a negotiated parameter, which spec §11 forbids.
    if (rs !== RECORD_SIZE || idlen !== 0) {
      throw new ContentError("header does not match the specification");
    }
    try {
      keys = await deriveContentKeys(fileKey, Uint8Array.from(salt));
    } catch {
      throw new ContentError("cannot derive content keys");
    }
    pending = pending.slice(HEADER_SIZE);
  };

  const decryptRecord = async (
    record: Uint8Array,
    controller: TransformStreamDefaultController<Uint8Array>,
  ) => {
    if (!keys) {
      throw new ContentError("header has not been read");
    }
    if (seq >= MAX_SEQUENCE || sawFinal) {
      throw new ContentError("data follows the final record");
    }

    let plaintext: Uint8Array;
    try {
      plaintext = new Uint8Array(
        await crypto.subtle.decrypt(
          {
            name: "AES-GCM",
            iv: recordNonce(keys.nonceBase, seq) as BufferSource,
            tagLength: 128,
          },
          keys.key,
          record as BufferSource,
        ),
      );
    } catch {
      throw new ContentError("record failed authentication");
    }
    if (plaintext.length === 0) {
      throw new ContentError("record has no delimiter");
    }

    // The delimiter is the final octet. Optional zero padding, which RFC 8188
    // permits, is deliberately not accepted: it would make two distinct
    // encodings of one plaintext valid.
    const delimiter = plaintext[plaintext.length - 1];
    if (delimiter === DELIMITER_FINAL) {
      sawFinal = true;
    } else if (delimiter === DELIMITER_NON_FINAL) {
      // A non-final record must be full; otherwise a truncated stream could be
      // presented as a complete one.
      if (record.length !== RECORD_SIZE) {
        throw new ContentError("short non-final record");
      }
    } else {
      throw new ContentError("invalid record delimiter");
    }

    seq++;
    if (plaintext.length > 1) {
      controller.enqueue(plaintext.subarray(0, plaintext.length - 1));
    }
  };

  return new TransformStream<Uint8Array, Uint8Array>({
    async transform(chunk, controller) {
      append(chunk);
      if (!keys) {
        if (pending.length < HEADER_SIZE) {
          return;
        }
        await readHeader();
      }
      while (pending.length >= RECORD_SIZE) {
        const record = pending.subarray(0, RECORD_SIZE);
        await decryptRecord(Uint8Array.from(record), controller);
        pending = pending.slice(RECORD_SIZE);
      }
    },
    async flush(controller) {
      if (!keys) {
        // Fewer than HEADER_SIZE bytes ever arrived.
        throw new ContentError("stream ended before the header");
      }
      if (pending.length > 0) {
        await decryptRecord(pending, controller);
        pending = new Uint8Array(0);
      }
      if (!sawFinal) {
        throw new ContentError("stream ended without a final record");
      }
    },
  });
}

/**
 * Drives a whole buffer through a transform and collects the output.
 *
 * When a transform errors, **both** sides of the stream reject. Each is
 * captured rather than awaited directly, so that propagating one does not leave
 * the other as an unhandled rejection. The readable side's error is preferred,
 * because it carries the cause; the writable side merely reports that the
 * stream was torn down.
 */
async function pump(
  input: Uint8Array,
  transform: TransformStream<Uint8Array, Uint8Array>,
): Promise<Uint8Array> {
  const writer = transform.writable.getWriter();
  const reader = transform.readable.getReader();

  const writing: Promise<unknown> = (async () => {
    if (input.length > 0) {
      await writer.write(input);
    }
    await writer.close();
  })().then(
    () => undefined,
    (error: unknown) => error ?? new Error("crypto: stream write failed"),
  );

  type ReadResult = { ok: true; value: Uint8Array } | { ok: false; error: unknown };
  const reading: Promise<ReadResult> = (async () => {
    const chunks: Uint8Array[] = [];
    for (;;) {
      const { done, value } = await reader.read();
      if (done) {
        break;
      }
      chunks.push(value);
    }
    const total = chunks.reduce((sum, c) => sum + c.length, 0);
    const out = new Uint8Array(total);
    let offset = 0;
    for (const c of chunks) {
      out.set(c, offset);
      offset += c.length;
    }
    return out;
  })().then(
    (value) => ({ ok: true as const, value }),
    (error: unknown) => ({ ok: false as const, error }),
  );

  const [writeError, readResult] = await Promise.all([writing, reading]);
  if (!readResult.ok) {
    throw readResult.error;
  }
  if (writeError !== undefined) {
    throw writeError;
  }
  return readResult.value;
}

/**
 * Encrypts a whole buffer. Convenient for small payloads and tests; use
 * {@link encryptStream} for anything that should not be held in memory.
 */
export function encryptBytes(
  fileKey: Uint8Array,
  plaintext: Uint8Array,
  contentSalt?: Uint8Array,
): Promise<Uint8Array> {
  return pump(plaintext, encryptStream(fileKey, contentSalt ?? randomBytes(CONTENT_SALT_SIZE)));
}

/** Decrypts a whole buffer produced by {@link encryptBytes}. */
export function decryptBytes(fileKey: Uint8Array, stream: Uint8Array): Promise<Uint8Array> {
  return pump(stream, decryptStream(fileKey));
}
